rep-0147.rst

```diff
@@ -0,0 +1,365 @@
+
+
+Reference Abstractions
+======================
+
+For this document we will use the following abstractions of the architecture.
+These abstractions are used to define the interfaces, however it is not required that a given system implement all the interfaces.
+A given system may implement subsets of these interfaces.
+
+
+Control Abstractions
+--------------------
+
+For controlling an aerial vehicle we will use the following terms for the interfaces.
+
+.. raw:: html
+  
+  <div class="mermaid">
+  graph LR
+    na[ ]
+    tc[Trajectory Controller]
+    ac[Attitude Controller]
+    rc[Rate Controller]
+    va[Vehicle Abstraction]
+    na ==>|Trajectory|tc
+    tc ==>|Pose|ac
+    ac ==>|Rate|rc
+    rc ==>|Acceleration|va
+    
+    va -.-> am[Allocation Matrix]
+    va -.-> mixer
+    
+    am -.-> Actuators
+    mixer -.-> Actuators 
+  </div>
+
+High Level Architecture
+-----------------------
+
+A higher level view of the system uses the following abstraction.
+
+.. raw:: html
+  
+  <div class="mermaid">
+  graph LR
+    GCS[Ground Control Station]
+    CS[Collision Sensors]
+    OS[Odometry Sensors]
+    C[Controller]
+    L[ ]
+    style L fill:#fff,stroke:#ff0,stroke-width:0px;
+    L --> CS
+    L --> GCS
+    L --> OS
+    subgraph Generic Planning Framework
+    CM[Collision Mapping]
+    MSC[Minimum Snap Controller]
+    CM ---|Shared State|MSC
+    end
+    CS -->|Obstacle Observations|CM
+    OS -->|Odometry|MSC
+    OS -->|Odometry|CM
+    OS -->|Odometry|C
+    GCS -->|Goals|MSC
+    MSC -->|Trajectory|C
+    %% Links to L 
+    linkStyle 0 fill:#fff,stroke:#ff0,stroke-width:0px;
+    linkStyle 1 fill:#fff,stroke:#ff0,stroke-width:0px;
+    linkStyle 2 fill:#fff,stroke:#ff0,stroke-width:0px;
+  </div>
+
+
+Coordinate Frames
+=================
+
+It is expected that users of this interface will also follow the coordinate frame conventions layed out in REP 105 [#rep105]_.
+
+Proposed Interfaces
+===================
+
+Control Interfaces
+------------------
+
+Trajectory Control Interface
+,,,,,,,,,,,,,,,,,,,,,,,,,,,,
+
+For sending trajectories in 3D space there already exists a message used for free space planning.
+This message used is ``trajectory_msgs/MultiDOFJointTrajectory`` [#multidoftraj]_ and is actively used by planning frameworks already.
+
+This message is not the most straight forward to parse as it has a lot of optional fields which increases the complexity and requires consistent labelling and filling.
+The complexity of the message parsing is traded off against supporting a relatively compact representations.
+
+* Datatype: ``trajectory_msgs/MultiDOFJointTrajectory`` [#multidoftraj]_ 
+* Topic: ``command_trajectory``
+
+Pose Interface
+,,,,,,,,,,,,,,
+
+The message to send a goal pose is ``geometry_msgs/PoseStamped`` [#posestamped]_.
+This is a very highly used standard message and contains precisely the information needed.
+
+* Datatype: ``geometry_msgs/PoseStamped`` [#posestamped]_
+* Topic: ``command_pose``
```

Comment:
Supporting a subset is definitely doable. Some of the more complicated messages such as the Trajectory messages do it with our closed approximation of optional fields which is empty arrays. The Pose message I see as the simplified interface to the single point trajectory message. 

In your examples. Goto X keeping Y the same. You just need to know the last command and only modify the values you want to change. You can get it by introspection, or even more likely your process sent the last command. It's important to keep things like these commands self contained so that the commands are not subject to hidden state in the system as well. 

For the example such as go to X height. As a position command that's completely underspecified. As a high level goal for a planner which will generate a trajectory to achieve flight at that height that's much more reasonable. And I think that supporting those higher level goals would be great, but I think that should be separate from the low level control. 

For sending relative commands it makes sense to support transforming them using the header frame_id. Note that the controller will need to transform them out of the frame of the vehicle into a world frame, otherwise move 2 meters ahead of the vehicle would never be achievable. Our standard at the moment is body x forward, y left, z up for vehicles. However like we do for cameras (x forward in default frame, 2nd frame published rotated with z forward with `_optical` suffix for the frameid) we could have a standard way to publish a 2nd rotated coordinate frame for convenience.

The header time stamp is a good question. For the trajectory messages there is a specific meaning. For the single point in time commands they represent the time at which the data should be interpreted and that the vehicle should be trying to achieve those goals immediately and continue to attempt to achieve them. Concretely if I wanted to hover 1m away from my current pose. I could send the command to hover at 1,0,0 at time t in the body frame. And as the vehicle moved forward the goal would be constant in position as it's relative to the position of the vehicle at time t, not t+n which is now.